R/training_run.R

```diff
@@ -125,32 +125,55 @@ training_run <- function(file = "train.R",
 #' @inheritParams ls_runs
 #'
 #' @param flags Named list with flag values (multiple values can be
-#'   provided for each flag)
+#'   provided for each flag).
+#' @param flag_grid A data frame that contains pre-generated
+#'   combinations of flags (e.g. via [base::expand.grid()]). This can
+#'   be useful for batch processing of large grids as well as
+#'   for subsetting combinations. See 'Examples'.
 #' @param sample Sampling rate for flag combinations (defaults to
 #'   running all combinations).
 #' @param confirm Confirm before executing tuning run.
 #'
 #' @return Data frame with summary of all training runs performed
 #'   during tuning.
 #'
-#' @examples \dontrun{
+#' @examples
+#' \dontrun{
 #' library(tfruns)
 #'
+#' # using the flags argument
 #' runs <- tuning_run("mnist_mlp.R", flags = list(
 #'   batch_size = c(64, 128),
 #'   dropout1 = c(0.2, 0.3, 0.4),
 #'   dropout2 = c(0.2, 0.3, 0.4)
 #' ))
-#'
 #' runs[order(runs$eval_acc, decreasing = TRUE), ]
 #'
+#' # using the flag_grid argument, but remove combinations
+#' # where the combined dropout rate exceeds 1
+#' grid <- expand.grid(
+#'   batch_size = 2^c(1:5),
+#'   dropout1 = seq(0.2, 0.8, by = 0.1),
+#'   dropout2 = seq(0.2, 0.8, by = 0.1)
+#' )
+#' grid$combined_droput <- grid$dropout1 + grid$dropout2
+#' grid$id <- 1:nrow(grid)
+#' grid <- grid[grid$combined_droput <= 1,]
+#' runs <- tuning_run("mnist_mlp.R", flag_grid = grid, sample = 0.01)
+#' # check if an arbitrary run has been completed
+#' completed_runs <- runs[runs$completed]$id
+#' 2 %in% completed_runs
+#' # continue evaluating the grid if the process was paused / interrupted etc.
+#' grid <- grid[!(grid$id %in% completed_runs), ]
+#' runs <- tuning_run("mnist_mlp.R", flag_grid = grid, sample = 0.01)
 #' }
 #'
 #' @export
 tuning_run <- function(file = "train.R",
                        context = "local",
                        config = Sys.getenv("R_CONFIG_ACTIVE", unset = "default"),
                        flags = NULL,
+                       flag_grid = NULL,
```

Comment:
What would you think of just overloading the `flags` argument to take a data frame as well (in which case we'd take the `flag_grid` code path)? I think that's consistent w/ the spirit of `flags` and would be one less parameter for people to grapple with in the documentation, etc.

Reply:
In general, I prefer type stability for inputs. However, I see your point since the API already has many arguments. I can adapt it if you prefer it that way.

Reply:
Yeah, in this case I think it's better we trade off on the type stability. If you could make that change it would great.

Reply:
I don't think we should have the example code be so complex/involved. Just the basic calling is all that's necessary (techniques for resuming, etc. might be better off in their own section or in a vignette).